### PR TITLE
Parquet: make row group check (min and max record count) configurable.

### PR DESCRIPTION
**Pull request by stevenzwu**
This is useufl when we want to use smaller row group size. In the past, we also found that tuning these configurations to smaller value is important lto avoid OOM problem when there are large records (like MBs). E.g., in the past, we have set the min/max check count default to 10/100 by default for the Flink streaming ingestion part. The performance impact is very insignificant. For streams with large records, we have even set the min/max to much smaller value like 1/5.

### CODE REVIEW

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java [hunk 1]
```diff
@@ -46,7 +46,7 @@
 
 class ParquetWriter<T> implements FileAppender<T>, Closeable {
 
-  private static DynConstructors.Ctor<PageWriteStore> pageStoreCtorParquet = DynConstructors
+  private static final DynConstructors.Ctor<PageWriteStore> pageStoreCtorParquet = DynConstructors
```
Comment (stevenzwu) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:49]: In this PR, I also fixed this compiler warnings and `RuntimeIOException` deprecation (to `UncheckedIOException`). Please let me know if it is preferred to leave those out as a separate PR

Comment (jackye1995) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java:49]: Thanks for working on that! I think we are accumulating more and more of these warnings, better to fix them as much as possible along the way.

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 2]
```diff
@@ -107,6 +107,18 @@ private TableProperties() {
...
+      "write.parquet.row-group-check-min-record-count";
+  public static final String DELETE_PARQUET_ROW_GROUP_CHECK_MIN_RECORD_COUNT =
+      "write.delete.parquet.row-group-check-min-record-count";
+  public static final String PARQUET_ROW_GROUP_CHECK_MIN_RECORD_COUNT_DEFAULT = "100";
+
+  public static final String PARQUET_ROW_GROUP_CHECK_MAX_RECORD_COUNT =
+      "write.parquet.row-group-check-max-record-count";
+  public static final String DELETE_PARQUET_ROW_GROUP_CHECK_MAX_RECORD_COUNT =
+      "write.delete.parquet.row-group-check-max-record-count";
+  public static final String PARQUET_ROW_GROUP_CHECK_MAX_RECORD_COUNT_DEFAULT = "10000";
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:120]: this can be integer

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 3]
```diff
@@ -107,6 +107,18 @@ private TableProperties() {
   public static final String DELETE_PARQUET_COMPRESSION_LEVEL = "write.delete.parquet.compression-level";
   public static final String PARQUET_COMPRESSION_LEVEL_DEFAULT = null;
 
+  public static final String PARQUET_ROW_GROUP_CHECK_MIN_RECORD_COUNT =
+      "write.parquet.row-group-check-min-record-count";
+  public static final String DELETE_PARQUET_ROW_GROUP_CHECK_MIN_RECORD_COUNT =
+      "write.delete.parquet.row-group-check-min-record-count";
+  public static final String PARQUET_ROW_GROUP_CHECK_MIN_RECORD_COUNT_DEFAULT = "100";
```
Comment (jackye1995) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:114]: this can be integer

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 4]
```diff
@@ -316,7 +331,13 @@ static Context dataContext(Map<String, String> config) {
 
         String compressionLevel = config.getOrDefault(PARQUET_COMPRESSION_LEVEL, PARQUET_COMPRESSION_LEVEL_DEFAULT);
 
-        return new Context(rowGroupSize, pageSize, dictionaryPageSize, codec, compressionLevel);
+        int rowGroupCheckMinRecordCount = Integer.parseInt(config.getOrDefault(
```
Comment (jackye1995) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:334]: can use `PropertyUtil.propertyAsInt`

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 5]
```diff
@@ -316,7 +331,13 @@ static Context dataContext(Map<String, String> config) {
...
+            PARQUET_ROW_GROUP_CHECK_MIN_RECORD_COUNT, PARQUET_ROW_GROUP_CHECK_MIN_RECORD_COUNT_DEFAULT));
+        int rowGroupCheckMaxRecordCount = Integer.parseInt(config.getOrDefault(
+            PARQUET_ROW_GROUP_CHECK_MAX_RECORD_COUNT, PARQUET_ROW_GROUP_CHECK_MAX_RECORD_COUNT_DEFAULT));
+
```
Comment (jackye1995) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:338]: I think we need to add some basic validations such as the config value is positive, max is greater than min.

**Review by jackye1995 — COMMENTED** [review]
Overall looks good to me. 

The only concern I have is that this seems to be a configuration specific to the use case on engine side. For example, in streaming we might want to turn this to smaller number, but for normal Spark ingestion the defaults might work better. 

But as of today we don't have a way to add runtime session options to the appender, it just takes table properties. I don't have enough historical context to say if that's intentional or not, maybe @rdblue can comment more about this. But overall I think it good that at least we have a config rather than hard-coded values.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 4 above)
Comment (stevenzwu) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: Originally, I was trying to stay with the existing code/style in this class. In the new commit, I have updated `Integer.partInt` to `PropertyUtil.propertyAsInt` for all integer configs (old or new)

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 5 above)
Comment (stevenzwu) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:351]: thx for pointing this out. added validations.

**Comment by stevenzwu** [comment]
@jackye1995 Please take another look.

Regarding your comment that new configs is more specific to the use case on engine side, I actually this is not engine specific. Sure, it probably matters a little more on the streaming ingestion (Flink or Spark streaming). It can matter to batch write too. 

E.g., We want to have smaller row group size (like 16 MB) to be able to split files into more splits for higher parallelism.  if the average row size is big (like MBs), then we need to tune down these configs to have more accurate control on the target row group size. This is irrespective to streaming or batch write.

**Comment by stevenzwu** [comment]
@rdblue can you help take a look?

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 6]
```diff
@@ -291,32 +301,56 @@ private WriteBuilder createContextFunc(Function<Map<String, String>, Context> ne
...
+        this.rowGroupCheckMaxRecordCount = rowGroupCheckMaxRecordCount;
       }
 
       static Context dataContext(Map<String, String> config) {
-        int rowGroupSize = Integer.parseInt(config.getOrDefault(
-            PARQUET_ROW_GROUP_SIZE_BYTES, PARQUET_ROW_GROUP_SIZE_BYTES_DEFAULT));
+        int rowGroupSize = PropertyUtil.propertyAsInt(config,
+            PARQUET_ROW_GROUP_SIZE_BYTES, PARQUET_ROW_GROUP_SIZE_BYTES_DEFAULT);
+        Preconditions.checkArgument(rowGroupSize > 0,
+            "Row group size must be > 0");
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:323]: Does this need to be on a separate line?

File: parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java [hunk 7]
```diff
@@ -51,8 +54,15 @@
...
-    File parquetFile = generateFileWithTwoRowGroups(null).first();
+    Map<String, String> props = ImmutableMap.of(
+        PARQUET_ROW_GROUP_SIZE_BYTES,
+        Integer.toString(4 * Integer.BYTES));
+    // Without an explicit writer function doesn't support PARQUET_ROW_GROUP_CHECK_MIN_RECORD_COUNT
+    // PARQUET_ROW_GROUP_CHECK_MAX_RECORD_COUNT configs.
+    // Even though row group size is 16 bytes, we still have to write 101 records
+    // as default PARQUET_ROW_GROUP_CHECK_MIN_RECORD_COUNT is 100.
+    File parquetFile = generateFileWithTwoRowGroups(null, 101, props)
+        .first();
```
Comment (rdblue) [inline parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java:65]: Does this need to be on a separate line?

File: parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java [hunk 8]
```diff
@@ -97,15 +118,14 @@ public void testNumberOfBytesWritten() throws IOException {
     Assert.assertEquals(expectedSize, actualSize);
   }
 
-  private Pair<File, Long> generateFileWithTwoRowGroups(Function<MessageType, ParquetValueWriter<?>> createWriterFunc)
+  private Pair<File, Long>  generateFileWithTwoRowGroups(
```
Comment (rdblue) [inline parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java:121]: The modification to these tests doesn't seem to fit. This is pulling the writer configuration out of this method, but still using the name `generateFileWithTwoRowGroups` even though there's no longer a guarantee that will actually happen. I think it would be better to build the properties in this method and pass in settings:

```java
private ... generateFile(Function<...> createWriterFunc, Long rowGroupSizeBytes, Integer minCheckRecordCount, Integer maxCheckRecordCount)
```

That way, all this function does is create the file. I think it should also set its own desired record count based on the min row group count so you don't have to pass in 1 more than the row group size in records.

**Review by rdblue — CHANGES_REQUESTED** [review]
The main changes look good, but I think this could make the test modifications more clear.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 6 above)
Comment (stevenzwu) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: fixed

File: parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java (hunk 7 above)
Comment (stevenzwu) [reply parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java]: fixed

File: parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java (hunk 8 above)
Comment (stevenzwu) [reply parquet/src/test/java/org/apache/iceberg/parquet/TestParquet.java]: Updated the method to build the props inside this method.

`desiredRecordCount` is already part of the method signature. No more hardcoded 1 more than the row group size in records.

**Review by kbendick — APPROVED** [review]
LGTM
